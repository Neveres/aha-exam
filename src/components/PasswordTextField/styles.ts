import { css } from '@emotion/react'

import { fontSize, fontWeight } from 'src/GlobalCss'

export const passwordInput = css`
  width: 335px;
  height: 58px;
  box-sizing: border-box;

  * {
    font-family: 'ubuntu' !important;
  }

  .MuiInputBase-root {
    height: 100%;

    input {
      position: absolute;
      top: 15px;
      width: 311px;
      height: 24px;
      padding: 0 12px;
      font-size: ${fontSize.large};
      font-weight: ${fontWeight.lighter};
      line-height: 24px;
      letter-spacing: 0.15px;
    }

    fieldset {
      height: 100%;
      box-sizing: border-box;
      padding: 0 0 0 9px;

      legend {
        width: 63px;
        height: 0;
        span {
          padding-left: 0;
        }
      }
    }
  }
`
