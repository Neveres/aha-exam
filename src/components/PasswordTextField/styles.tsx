import { css } from '@emotion/react'

export const passwordInput = css`
  margin-top: 50px;
`
