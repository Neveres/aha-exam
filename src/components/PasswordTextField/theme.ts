import { createTheme } from '@mui/material'

export const theme = createTheme({
  components: {
    MuiOutlinedInput: {
      styleOverrides: {
        root: {
          borderRadius: 10,
          '&:hover .MuiOutlinedInput-notchedOutline': {
            borderColor: 'white',
          },
          '& .MuiOutlinedInput-notchedOutline': {
            border: '3px solid #888888',
          },
          '&.Mui-focused': {
            '& .MuiOutlinedInput-notchedOutline': {
              border: '3px solid #0584CC',
            },
          },
        },
      },
    },
  },
})
