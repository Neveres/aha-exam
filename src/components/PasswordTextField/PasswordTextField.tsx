/** @jsxImportSource @emotion/react */
import React, { useCallback, useMemo, useState } from 'react'
import { TextField, ThemeProvider } from '@mui/material'

import { colors } from 'src/GlobalCss'
import { validatePassword } from 'src/libraries'
import { PurePasswordValidation } from 'src/components'
import { theme } from './theme'
import { passwordInput } from './styles'

export const PasswordTextField = () => {
  const symbolOfInput = '*'
  const [password, setPassword] = useState('')

  const onChange = useCallback(
    (event: React.ChangeEvent<HTMLInputElement>) => {
      const {
        target: { value },
      } = event

      const { length: lengthOfPassword } = password
      setPassword(
        value.length > lengthOfPassword
          ? `${password}${value.slice(-1)}`
          : password.slice(0, lengthOfPassword - 1),
      )
    },
    [password],
  )

  const validationResult = useMemo(() => validatePassword(password), [password])
  const encryptedPassword = useMemo(
    () => symbolOfInput.repeat(password.length),
    [password],
  )

  return (
    <>
      <ThemeProvider theme={theme}>
        <TextField
          label="Password"
          InputLabelProps={{
            shrink: true,
            style: {
              transform: 'translate(14px, -9px) scale(1)',
              fontSize: '12px',
              color: colors.label,
              lineHeight: '18px',
              letterSpacing: '0.4px',
              textOverflow: 'unset',
              left: '2px',
              top: '-3px',
            },
          }}
          inputProps={{
            style: { color: colors.label },
          }}
          placeholder="Password"
          css={passwordInput}
          value={encryptedPassword}
          onChange={onChange}
        />
      </ThemeProvider>
      {password && (
        <PurePasswordValidation validationResult={validationResult} />
      )}
    </>
  )
}
