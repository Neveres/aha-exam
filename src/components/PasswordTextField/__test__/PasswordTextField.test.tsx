import React from 'react'
import { create } from 'react-test-renderer'

import { PasswordTextField } from '../PasswordTextField'

describe('PasswordTextField', () => {
  test('should render well', () => {
    expect(create(<PasswordTextField />).toJSON()).toMatchSnapshot()
  })
})
