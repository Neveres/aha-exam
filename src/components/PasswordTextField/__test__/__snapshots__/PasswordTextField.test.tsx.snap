// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordTextField should render well 1`] = `
<div
  className="MuiFormControl-root MuiTextField-root css-hb9hc9-MuiFormControl-root-MuiTextField-root-PasswordTextField"
>
  <label
    className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
    data-shrink={true}
    htmlFor=":r0:"
    id=":r0:-label"
    style={
      Object {
        "color": "#ffffff",
        "fontSize": "12px",
        "left": "1px",
        "letterSpacing": "0.4px",
        "lineHeight": "18px",
        "textOverflow": "unset",
        "transform": "translate(14px, -9px) scale(1)",
      }
    }
  >
    Password
  </label>
  <div
    className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-e2jvzi-MuiInputBase-root-MuiOutlinedInput-root"
    onClick={[Function]}
  >
    <input
      aria-invalid={false}
      autoFocus={false}
      className="MuiInputBase-input MuiOutlinedInput-input css-24rejj-MuiInputBase-input-MuiOutlinedInput-input"
      disabled={false}
      id=":r0:"
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Password"
      required={false}
      style={
        Object {
          "color": "#ffffff",
        }
      }
      type="text"
      value=""
    />
    <fieldset
      aria-hidden={true}
      className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        className="css-1in441m"
      >
        <span>
          Password
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;
